### PR TITLE
Torchx mcad coscheduler

### PR DESCRIPTION
**Pull request by Sara-KS**
This addition enables the option of running TorchX-MCAD with a Kubernetes co-scheduler and PodGroups. In shared Kubernetes clusters where some users do not represent their jobs with AppWrappers, the additional use of a co-scheduler helps ensure correct gang scheduling. 

Test plan:
Updated CI test coverage
Testing includes using the new`--coscheduler_name` flag for the kubernetes_mcad scheduler and ensuring that Kubernetes or OpenShift reports the named scheduler when pods are scheduled.

### CODE REVIEW

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 1]
```diff
@@ -352,6 +353,31 @@ def role_to_pod(
     )
 
 
+def create_pod_group(role: Role, namespace: str, app_id: str) -> "Dict[str, Any]":
+    pod_group_name = app_id + "-" + cleanup_str(role.name) + "-pg"
+
+    pod_group: Dict[str, Any] = {
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:359]: Would be nice to add a  link to the schema definition for this type. I think it's https://github.com/kubernetes-sigs/scheduler-plugins/blob/56e2398e5051f796d5f02adb821d57eb8b6c20a7/config/crd/bases/scheduling.sigs.k8s.io_podgroups.yaml#L9 ?

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 2]
```diff
@@ -473,8 +506,13 @@ def app_to_resource(
                 replica_role,
                 service_account,
                 image_secret,
+                coscheduler_name,
+            )
+            pod.metadata.labels.update(
+                pod_labels(
+                    app, role_idx, role, replica_id, coscheduler_name, unique_app_id
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:513]: there's a lot of fields here. Think it's better to switch these to use `app=app, coscheduler_name=coscheduler_name, ...` to avoid accidentally passing the wrong field to the wrong param

File: torchx/schedulers/kubernetes_mcad_scheduler.py [hunk 3]
```diff
@@ -861,9 +900,15 @@ def _submit_dryrun(
...
         assert isinstance(namespace, str), "namespace must be a str"
 
+        coscheduler_name = cfg.get("coscheduler_name")
+        assert coscheduler_name is None or isinstance(
+            coscheduler_name, str
+        ), "coscheduler_name must be a string"
+
         resource = app_to_resource(
-            app, namespace, service_account, image_secret, priority
+            app, namespace, service_account, image_secret, coscheduler_name, priority
```
Comment (d4l3k) [inline torchx/schedulers/kubernetes_mcad_scheduler.py:909]: likewise kwargs for this

**Review by d4l3k — APPROVED** [review]

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 1 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py:359]: Included the PodGroup CRD (release 1.24) in the updated documentation.

File: torchx/schedulers/kubernetes_mcad_scheduler.py (hunk 3 above)
Comment (Sara-KS) [reply torchx/schedulers/kubernetes_mcad_scheduler.py]: Updated

**Comment by kiukchung** [comment]
> 2/27/2023 Component Integration Tests/ Kubernetes Dist Train Integration Tests appear to be failing due to a missing environment variable: torchx/scritps/integ_test_utils.py expects `CONTAINER_REPO`. Current test failures indicate that the container repository information is not detected in the test environment and attempts a docker push with ":tag" format instead of "repo:tag". Related issue: #694

Hi @Sara-KS thanks for taking a look I actually fixed that particular issue yesterday (see: https://github.com/pytorch/torchx/blob/main/scripts/kube_dist_trainer.py#L45) but the `Kubernetest Dist Train Integration Test` is failing for a different reason (401 Unauthorized when querying for the status of the job): https://github.com/pytorch/torchx/actions/runs/4277421091/jobs/7446219941

Since I've left Meta and don't have access to TorchX's AWS account I'm unable to login and take a deeper look. Perhaps @kurman or @priyaramani  can help here.

Long term, @d4l3k and I have discussed moving that integ test to run directly on the CI runner with minikube instead of hitting a deployed EKS cluster. Tristan's already made `components integ test` and `kfp integ test` to do this, we just haven't gotten around making the same change for the k8s integ test.

FWIW `Components Integration Tests` should be passing now (if you rebase on top of main)

**Comment by d4l3k** [comment]
That integration test doesn't use the mcad scheduler anyways so you can just ignore it for this PR

**Comment by d4l3k** [comment]
LGTM Failures look like an oversight on my part on how tagging is handled for forks
